src/transformers/modeling_bart.py

```diff
@@ -935,12 +943,269 @@ def forward(
+            model_inputs = self.prepare_inputs_for_generation(
+                input_ids, decoder_cache, decoder_input_ids, attention_mask,
+            )
+            outputs = self(**model_inputs)
+            lprobs = F.log_softmax(outputs[0][:, -1, :], dim=-1)
+
+            assert pad_token_id is not None, "should break circleci"
+            lprobs[lprobs != lprobs] = -math.inf  # kill nans
+            lprobs[:, pad_token_id] = -math.inf
+            # TODO(SS): unk is 3? fairseq also takes out every step, and has unk at slot 3
+
+            assert bos_token_id == 0, "configurable bos_token_id not yet supported"
+            if step == max_length:  # FORCE EOS to be chosen
+                lprobs[:, :eos] = -math.inf
+                lprobs[:, eos + 1 :] = -math.inf
+            elif step == 0:  # Force BOS to be chosen
+                lprobs[:, bos_token_id + 1 :] = -math.inf
+            if step < min_len:
+                lprobs[:, eos] = -math.inf
+            assert self._do_output_past(outputs)
+            decoder_cache = outputs[1]
+            if repetition_penalty != 1.0:
+                self.enforce_repetition_penalty_(lprobs, batch_size, num_beams, prev_output_tokens, repetition_penalty)
+            num_hypos = batch_size * num_beams
+            if no_repeat_ngram_size > 0:  # copied from fairseq
+                # for each beam and batch sentence, generate a list of previous ngrams
+                banned_tokens = self.calc_banned_tokens(prev_output_tokens, num_hypos, no_repeat_ngram_size, step)
+                # then set their probabilities to -inf
+                for bbsz_idx in range(num_hypos):
+                    lprobs[bbsz_idx, banned_tokens[bbsz_idx]] = -math.inf
+            assert lprobs.size() == (batch_size * num_beams, vocab_size)
+            _scores = lprobs + beam_scores[:, None].expand_as(lprobs)  # (batch_size * num_beams, vocab_size)
+
+            # re-organize to group the beam together (we are keeping top hypothesis across beams)
+            _scores = _scores.view(batch_size, num_beams * vocab_size)  # (batch_size, num_beams * vocab_size)
+            next_scores, next_words = torch.topk(_scores, 2 * num_beams)
+            assert next_scores.size() == next_words.size() == (batch_size, 2 * num_beams)
+
+            # list of (batch_size * num_beams)
+            next_batch_beam = []  # Tuple(next score, next word, current position in the batch)
+            for batch_idx in range(batch_size):
+                # if we are done with this sentence (because we can't improve)
+                if done[batch_idx]:  # then pad all associated hypotheses
+                    assert (
+                        len(finalized_hyps[batch_idx]) >= num_beams
+                    ), "Example can only be done if at least {} beams have been generated".format(num_beams)
+                    next_batch_beam.extend([(0, pad_token_id, 0)] * num_beams)  # pad the batch
```

Comment:
why can we assume that the beam_idx is 0 here? @thomwolf

Reply:
I think this is taken from here: 
https://github.com/facebookresearch/XLM/blob/9e6f6814d17be4fe5b15f2e6c43eb2b2d76daeb4/src/model/transformer.py#L625

Once one `batch_idx` is done it will always be done afterwards and we don't really care about the `beam_idx` anymore. So I think the `beam_idx` equals 0 here out of convenience and because no matter what the `beam_idx` is from this point onwards only the `pad_token_id` will be appended to all beams of this `batch_idx`